src/System.Threading.Channels/tests/Stress.cs

```diff
@@ -0,0 +1,423 @@
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSingleReaderWriterOptions), UnbufferedSingleReaderWriterOptions };
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSingleWriterOptions), UnbufferedSingleWriterOptions };
+
+            //
+            // Unbuffered Channel (AllowSynchronousContinuations)
+            //
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousOptions), UnbufferedSynchronousOptions };
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousSingleReaderOptions), UnbufferedSynchronousSingleReaderOptions };
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousSingleReaderWriterOptions), UnbufferedSynchronousSingleReaderWriterOptions };
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousSingleWriterOptions), UnbufferedSynchronousSingleWriterOptions };
+            yield return new object[] { Channel.CreateUnbuffered<int>(UnbufferedSynchronousOptions), UnbufferedSynchronousOptions };
+        }
+
+        const int MAX_NUMBER_TO_WRITE_TO_CHANNEL = 1_000_000;
+        private static readonly int MaxTaskCounts = Math.Max(2, Environment.ProcessorCount);
+
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void RunInStressMode(Channel<int> channel, ChannelOptions options)
+        {
+            ChannelReader<int> reader = channel.Reader;
+            ChannelWriter<int> writer = channel.Writer;
+            BoundedChannelOptions boundedOptions = options as BoundedChannelOptions;
+            bool shouldReadAllWrittenValues = boundedOptions == null || boundedOptions.FullMode == BoundedChannelFullMode.Wait;
+
+            List<Task> taskList = new List<Task>();
+
+            int readerTasksCount;
+            int writerTasksCount;
+
+            if (options.SingleReader)
+            {
+                readerTasksCount = 1;
+                writerTasksCount = options.SingleWriter ? 1 : MaxTaskCounts - 1;
+            }
+            else if (options.SingleWriter)
+            {
+                writerTasksCount = 1;
+                readerTasksCount = MaxTaskCounts - 1;
+            }
+            else
+            {
+                readerTasksCount = MaxTaskCounts / 2;
+                writerTasksCount = MaxTaskCounts - readerTasksCount;
+            }
+
+            int readCount = 0;
+            int wroteCount = 0;
+
+            for (int i=0; i < readerTasksCount; i++)
+            {
+                taskList.Add(new Task(() => {
+                    while (true)
+                    {
+                        try
+                        {
+                            if (!reader.TryRead(out int value))
+                            {
+                                value = reader.ReadAsync().GetAwaiter().GetResult();
```

Comment:
I noticed when I was using wait, the task was signaled (in Task.WaitAll) even the whole reading/writing is not executed yet. this is why I preferred the blocking. I'll take a look again at this.

Reply:
That's because you were using Task's ctor, which used in this way would take an Action, and that Action would return to the Task ctor on the first await.
https://blogs.msdn.microsoft.com/pfxteam/2012/02/08/potential-pitfalls-to-avoid-when-passing-around-async-lambdas/
That won't be an issue when you switch to use Task.Run.